libs/src/services/creatorNode/CreatorNodeSelection.js

```diff
@@ -0,0 +1,153 @@
+  constructor ({ creatorNode, numberOfNodes, ethContracts, whitelist, blacklist }) {
+    super({
+      getServices: async () => {
+        this.currentVersion = await ethContracts.getCurrentVersion(CREATOR_NODE_SERVICE_NAME)
+        const services = await this.ethContracts.getServiceProviderList(CREATOR_NODE_SERVICE_NAME)
+        return services.map(e => e.endpoint)
+      },
+      whitelist
+    })
+    this.creatorNode = creatorNode
+    this.numberOfNodes = numberOfNodes
+    this.ethContracts = ethContracts
+    this.blacklist = blacklist
+
+    // List of valid past creator node versions registered on chain
+    this.validVersions = null
+  }
+
+  /**
+   * Selects a primary and secondary creator nodes
+   *
+   * 1. Retrieve all the creator node services
+   * 2. Filter from/out creator nodes based off of the whitelist and blacklist
+   * 3. Init a map of default stats of each service
+   * 4. Perform a health check and sync status check to determine health and update stats
+   * 5. Sort by healthiest (highest version -> lowest version); secondary check if equal version based off of responseTime)
+   * 6. Select a primary and numberOfNodes-1 number of secondaries
+   */
+  async select () {
+    // Reset decision tree
+    this.decisionTree = []
+
+    // Get all the creator node endpoints on chain and filter
+    let services = await this.getServices()
+    this.decisionTree.push({ stage: 'Get All Services', val: services })
+
+    if (this.whitelist && this.whitelist.length > 0) {
+      services = this.filterToWhitelist(services)
+      this.decisionTree.push({ stage: 'Filtered To Whitelist', val: services })
+    }
+
+    if (this.blacklist && this.blacklist.length > 0) {
+      services = this.filterFromBlacklist(services)
+      this.decisionTree.push({ stage: 'Filtered From Blacklist', val: services })
+    }
+
+    // Init map of creator node endpoints and their default stats
+    const map = {}
+    services.map(service => {
+      map[service] = {
+        version: null,
+        healthCheckEndpoint: `${service}/health_check`,
+        healthCheckResponseTime: null,
+        isHealthy: false,
+        isBehind: true,
+        isConfigured: false,
+        geographicData: {}
+        // todo: add other criteria we should evaluate like diskUsage, currentNumberOfRequests
+      }
+    })
+
+    const healthyEndpoints = []
+    const healthyServices = []
+    for (const service of services) {
+      const start = Date.now()
+      try {
+        // Check health and sync status, and then update stats per service
+        let healthCheckResp = await axios({
+          method: 'get',
+          url: map[service].healthCheckEndpoint
+        })
+        healthCheckResp = healthCheckResp.data.data
+
+        map[service].healthCheckResponseTime = Date.now() - start
+        map[service].version = healthCheckResp.version
+        const { isBehind, isConfigured } = await this.creatorNode.getSyncStatus(service)
```

Comment:
what's this intended to do? is this for users that need their node sets reselected? for first time non creators this will return nothing